### PR TITLE
Add the shift-click selection interval in XY chart

### LINKED ISSUE #439: The shift-click selection interval in XY charts does not work.
When i want to enlarge or reduce my selection with shift click it doesn't work in xy charts.

**Here is how it should work :**

shift click

### PR DESCRIPTION
**Pull request by IbrahimFradj**
We can select interval with shift-click.

**Here is the result :**

shiftclick

Fixes #439.

Signed-off-by: Ibrahim Fradj <ibrahim.fradj@ericsson.com>

### CODE REVIEW

**Comment by ebugden** [comment]
Is this a feature in Trace Compass that current TC users expect to have and enjoy? 

If not, could you explain:
* An example where it's much more convenient to adjust an existing selected area instead of re-selecting the adjusted smaller/larger area?
* Why implementing adjusting an existing area with shift-click is the most practical/intuitive? (ex. Instead of being able to click on the bars showing the start/end of the selected area to drag them smaller/larger)

**Comment by IbrahimFradj** [comment]
> 
> 
> I definitely should have asked the following questions before assigning the associated issue to "High priority". I apologize for not asking these earlier!
> 
> Is this a feature in Trace Compass that current TC users expect to have and enjoy?
> 
> If not, could you explain:
> 
>     * An example where it's much more convenient to adjust an existing selected area instead of re-selecting the adjusted smaller/larger area?
> 
>     * Why implementing adjusting an existing area with shift-click is the most practical/intuitive? (ex. Instead of being able to click on the bars showing the start/end of the selected area to drag them smaller/larger)
> 
> 
> Thanks for working on the XY chart navigation! ☀️

In fact, this functionality exists in Trace compass and in timeline chart on theia-trace-extension. my goal is that xy chart has the same functionality as timeline chart.
This can be useful when for example I find a place where I want to make a selection without changing the beginning of the selection

**Comment by bhufmann** [comment]
> I definitely should have asked the following questions before assigning the associated issue to "High priority". I apologize for not asking these earlier!
> 
> Is this a feature in Trace Compass that current TC users expect to have and enjoy?
> 
> If not, could you explain:
> 
>     * An example where it's much more convenient to adjust an existing selected area instead of re-selecting the adjusted smaller/larger area?
> 
>     * Why implementing adjusting an existing area with shift-click is the most practical/intuitive? (ex. Instead of being able to click on the bars showing the start/end of the selected area to drag them smaller/larger)
> 
> 
> Thanks for working on the XY chart navigation! sunny

The advantage of this capability is that users can keep one side of selection (start or end) constant and then extend the selection without changing the starting point to measure time from a given timestamp on.

**Comment by ebugden** [comment]
@IbrahimFradj Can you rebase on master so that you get the recent Gitpod fixes? I'd like to play around with this feature to see if there's a quick way we could make it discoverable (ex. changing the cursor to a ⟷ when they press shift while hovering on a graph).

Thank you for the extra context IbrahimFradj and bhufmann. I don't know the practical analysis applications of measuring times with a fixed reference, but I trust your experience.

**Comment by IbrahimFradj** [comment]
> 
> 
> @IbrahimFradj Can you rebase on master so that you get the recent Gitpod fixes? I'd like to play around with this feature to see if there's a quick way we could make it discoverable (ex. changing the cursor to a ⟷ when they press shif

**Comment by IbrahimFradj** [comment]
> 
> 
> @IbrahimFradj Can you rebase on master so that you get the recent Gitpod fixes? I'd like to play around with this feature to see if there's a quick way we could make it discoverable (ex. changing the cursor to a ⟷ when they press shift while hovering on a graph).
> 
> Thank you for the extra context **IbrahimFradj** and **bhufmann**. I don't know the practical analysis applications of measuring times with a fixed reference, but I trust your experience.

@ebugden  I rebased on master, gitpod works.

**Comment by PatrickTasse** [comment]
> Why implementing adjusting an existing area with shift-click is the most practical/intuitive? (ex. Instead of being able to click on the bars showing the start/end of the selected area to drag them smaller/larger)

This could make it impossible to replace a narrow range selection with a new single selection in one click (without needing to click somewhere far away first).

Shift to extend selection also applies when using keyboard left/right navigation. That is really useful to make a selection to the exact time bounds of a state or group of states.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 1]
```diff
@@ -45,14 +45,11 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
     private updateSelection = (event: MouseEvent) => {
-        const scale = this.props.viewRange.getEnd() - this.props.viewRange.getstart();
         if (this.mouseIsDown && this.props.unitController.selectionRange && !this.isRightClick) {
             const xStartPos = this.props.unitController.selectionRange.start;
-            let end = xStartPos + ((event.screenX - this.posPixelSelect) / this.lineChartRef.current.chartInstance.width) * scale;
-            end = Math.min(Math.max(end, 0), this.props.unitController.absoluteRange);
             this.props.unitController.selectionRange = {
                 start: xStartPos,
-                end: end
+                end: this.getTimeX(event.offsetX)
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:52]: If you drag selection outside of the bounds of the XY chart then you start receiving event.offsetX in another component's coordinates and the end time is wrong.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

**Review by PatrickTasse — APPROVED** [review]

**Comment by ebugden** [comment]
> I rebased on master, gitpod works.

Looks good to me! Thanks for the update. Consistent navigation features across chart types is important to avoid user confusion.

**Review by bhufmann — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**